File: DFS_BFS/543-nearestExit-IDS.py

def nearestExit(m,entry):
    rows=len(m)
    cols=len(m[0])
    
    def isExit(r,c):
        if r==0 or c==0 or r==rows-1 or c==cols-1:
            return m[r][c]=='.'
        return False
    
    def DLS(pos,d):
        if isExit(pos[0],pos[1]) and pos!=tuple(entry):
            return True
        # if pos in seen:
        #     return False
        if d<=0:
            return False
        # seen.add(pos)
        r,c=pos
        m[r][c]='+'
        for dr,dc in ((0,1),(1,0),(-1,0),(0,-1)):
            nr,nc=dr+r,dc+c
            # r,c=pos
            #nr,nc=dr+r,dc+c
            if 0<=nr<=rows-1 and 0<=nc<=cols-1 and m[nr][nc]=='.' :
                if DLS((nr,nc),d-1):
                    return True
        # seen.remove(pos)
        m[r][c]='.'
        return False
            
    d=0
    D=rows*cols
    seen=set()
    while d<D:
        if DLS(tuple(entry),d):
            return d
        d+=1
    return -1
'''    
class Solution:
    def nearestExit(self, maze,entrance): 
        rows = len(maze)
        assert rows > 0
        cols = len(maze[0])
        assert cols > 0
        
        # @cache
        def isExit(r, c):
            if r == 0 or c == 0 or r == rows - 1 or c == cols - 1:
                return maze[r][c] == '.'
            return False

        def dls(r, c, d):
            if (r, c) != tuple(entrance) and isExit(r, c):
                return True
            if d <= 0:
                return False
            maze[r][c] = "+"
            for dr, dc in ((0, 1), (1, 0), (0, -1), (-1, 0)):
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and maze[nr][nc] == '.':                    
                    if dls(nr, nc, d - 1):
                        return True
            maze[r][c] = "."
            return False

        d = 0
        D = rows * cols
        while d < D:
            if dls(entrance[0], entrance[1], d):
                return d
            d += 1
        
        return -1
'''
maze = [["+","+",".","+"],[".",".",".","+"],["+","+","+","."]]
entrance = [1,2]

print(nearestExit(maze, entrance))

File: DFS_BFS/543-nearestExit-IDS.cpp
#include<iostream>
#include<set>
#include<vector>

using std::cout;
using VEC=std::vector<char>;
using VEC_INT=std::vector<int>;
using MATRIX=std::vector<VEC>;


bool isExit(int r,int c,const MATRIX& m)
{
    int rows=m.size();
    int cols=m[0].size();
    if(r==0||c==0 ||r==rows-1||c==cols-1)
        return m[r][c]=='.';
    return false;
}


bool DLS(MATRIX& m,int r,int c,int d,const VEC_INT& entry)
{
    int rows=m.size();
    int cols=m[0].size();

    if(VEC_INT(r,c)!=entry && isExit(r,c,m))
        return true;

    if(d<=0)
        return false;
    m[r][c]='+';
    std::vector<VEC_INT> dirs{{1,0},{0,1},{-1,0},{0,-1}};

    for(auto &dir:dirs)
    {
        int nr=r+dir[0];
        int nc=c+dir[1];
        if(nr>=0 &&nr<rows &&nc>=0 &&nc<cols &&m[nr][nc]=='.')
        {
            if( DLS(m,nr,nc,d-1,entry))
                return true;
        }
    }
    m[r][c]='.';
}

int nearestExit( MATRIX& m,  const VEC_INT& entry)
{
    int d=0;
    int D=m[0].size()*m.size();
    while(d<D)
    {
        if(DLS(m,entry[0],entry[1],d,entry))
            return d;
        d+=1;
    }
    return -1;
}


int main()
{
        // maze = [['+','+','.','+'],['.','.','.','+'],['+','+','+','.']]
// entrance = [1,2]

    MATRIX m{{'+','+','.','+'},{'.','.','.','+'},{'+','+','+','.'}};
    VEC_INT entrance{1,2};
    cout<<nearestExit(m,entrance);
}
